File: typescript/src/detect.ts
// Framework auto-detection — inspect message shape to determine the source SDK.

/**
 * Detect which framework produced the given messages.
 *
 * Uses duck-typing heuristics — no framework imports required.
 *
 * Detection order (first match wins):
 * 1. MemoSiftMessage instances → "memosift"
 * 2. Dict with "function_calls" or "function_responses" → "adk"
 * 3. Object with .additional_kwargs property → "langchain"
 * 4. Object with Agent SDK type names → "agent_sdk"
 * 5. Dict with "content" as array of typed blocks → "anthropic"
 * 6. Default → "openai"
 */
export function detectFramework(
  messages: readonly unknown[],
): "openai" | "anthropic" | "agent_sdk" | "adk" | "langchain" | "memosift" | "vercel_ai" {
  if (messages.length === 0) {
    throw new Error("Cannot detect framework from empty message list");
  }

  const samples = messages.slice(0, 5).filter((m) => m != null);
  if (samples.length === 0) return "openai";

  for (const msg of samples) {
    if (typeof msg !== "object" || msg === null) continue;
    const rec = msg as Record<string, unknown>;

    // 1. MemoSiftMessage — has _memosiftCompressed field.
    if ("_memosiftCompressed" in rec && "role" in rec && "content" in rec) {
      return "memosift";
    }

    // 2. Google ADK — uses function_calls/function_responses.
    if ("function_calls" in rec || "function_responses" in rec) {
      return "adk";
    }
    const parts = rec.parts;
    if (Array.isArray(parts)) {
      for (const part of parts) {
        if (
          typeof part === "object" &&
          part !== null &&
          ("function_call" in part || "function_response" in part)
        ) {
          return "adk";
        }
      }
    }

    // 3. LangChain — has additional_kwargs property.
    // Check before Agent SDK because LangChain's SystemMessage shares a name
    // with Agent SDK's SystemMessage, but has additional_kwargs.
    if ("additional_kwargs" in rec) {
      return "langchain";
    }

    // 4. Claude Agent SDK — typed objects with known class names.
    const typeName = (msg as { constructor?: { name?: string } }).constructor?.name;
    if (
      typeName !== undefined &&
      ["SystemMessage", "AssistantMessage", "UserMessage", "ResultMessage"].includes(typeName) &&
      typeName !== "Object"
    ) {
      return "agent_sdk";
    }

    // 5. Content as array of blocks — distinguish Vercel AI SDK from Anthropic.
    const content = rec.content;
    if (Array.isArray(content) && content.length > 0) {
      const firstBlock = content[0];
      if (typeof firstBlock === "object" && firstBlock !== null && "type" in firstBlock) {
        const blockType = (firstBlock as Record<string, unknown>).type;
        // 5a. Vercel AI SDK — uses "tool-call" and "tool-result" (hyphenated).
        if (blockType === "tool-call" || blockType === "tool-result") {
          return "vercel_ai";
        }
        // Also detect Vercel from toolCallId/toolName on parts (unique to Vercel).
        if (
          (blockType === "text" || blockType === "tool-call" || blockType === "tool-result") &&
          ("toolCallId" in firstBlock || "toolName" in firstBlock)
        ) {
          return "vercel_ai";
        }
        // 5b. Anthropic — uses "tool_use", "tool_result" (underscored).
        if (
          blockType === "text" ||
          blockType === "tool_use" ||
          blockType === "tool_result" ||
          blockType === "thinking" ||
          blockType === "image"
        ) {
          return "anthropic";
        }
      }
    }
  }

  // 6. Default — OpenAI.
  return "openai";
}

/** Valid framework identifiers. */
export const VALID_FRAMEWORKS = new Set([
  "openai",
  "anthropic",
  "agent_sdk",
  "adk",
  "langchain",
  "memosift",
  "vercel_ai",
] as const);

export type Framework =
  | "openai"
  | "anthropic"
  | "agent_sdk"
  | "adk"
  | "langchain"
  | "memosift"
  | "vercel_ai";

File: python/src/memosift/detect.py
# Framework auto-detection — inspect message shape to determine the source SDK.
from __future__ import annotations

from typing import Any


def detect_framework(messages: list[Any]) -> str:
    """Detect which framework produced the given messages.

    Inspects the shape of the first few messages to determine the source SDK.
    Uses duck-typing heuristics — no framework imports required.

    Detection order (first match wins):
    1. MemoSiftMessage instances → "memosift"
    2. Dict with "function_calls" or "function_responses" → "adk"
    3. Object with .additional_kwargs attribute → "langchain"
    4. Object with Agent SDK type names → "agent_sdk"
    5. Dict with "content" as a list of blocks → "anthropic"
    6. Default → "openai" (most permissive format)

    Args:
        messages: List of messages in any supported framework format.

    Returns:
        Framework identifier string.

    Raises:
        ValueError: If messages is empty.
    """
    if not messages:
        raise ValueError("Cannot detect framework from empty message list")

    # Sample first few non-None messages for detection.
    samples = [m for m in messages[:5] if m is not None]
    if not samples:
        return "openai"

    for msg in samples:
        # 1. MemoSiftMessage instances (check by attribute, not import).
        if (
            hasattr(msg, "_memosift_compressed")
            and hasattr(msg, "role")
            and hasattr(msg, "content")
        ):
            return "memosift"

        # 2. Google ADK — uses function_calls/function_responses instead of tool_calls.
        if isinstance(msg, dict):
            if "function_calls" in msg or "function_responses" in msg:
                return "adk"
            # Check nested in "parts" (ADK event structure).
            parts = msg.get("parts", [])
            if isinstance(parts, list):
                for part in parts:
                    if isinstance(part, dict) and (
                        "function_call" in part or "function_response" in part
                    ):
                        return "adk"

        # 3. LangChain — has additional_kwargs attribute or key (LangChain BaseMessage).
        # Check before Agent SDK because LangChain's SystemMessage shares a name
        # with Agent SDK's SystemMessage, but has additional_kwargs.
        if hasattr(msg, "additional_kwargs") and not isinstance(msg, dict):
            return "langchain"
        if isinstance(msg, dict) and "additional_kwargs" in msg:
            return "langchain"

        # 4. Claude Agent SDK — typed message objects.
        type_name = type(msg).__name__
        if type_name in ("SystemMessage", "AssistantMessage", "UserMessage", "ResultMessage"):
            module = type(msg).__module__ or ""
            if "langchain" not in module:
                return "agent_sdk"

        # 5. Content as list of blocks — distinguish Vercel AI SDK from Anthropic.
        if isinstance(msg, dict):
            content = msg.get("content")
            if isinstance(content, list) and len(content) > 0:
                first_block = content[0]
                if isinstance(first_block, dict):
                    block_type = first_block.get("type", "")
                    # 5a. Vercel AI SDK — uses "tool-call" and "tool-result" (hyphenated).
                    if block_type in ("tool-call", "tool-result"):
                        return "vercel_ai"
                    if block_type in ("text", "tool-call", "tool-result") and (
                        "toolCallId" in first_block or "toolName" in first_block
                    ):
                        return "vercel_ai"
                    # 5b. Anthropic — uses "tool_use", "tool_result" (underscored).
                    if block_type in (
                        "text",
                        "tool_use",
                        "tool_result",
                        "thinking",
                        "image",
                    ):
                        return "anthropic"

    # 6. Default — OpenAI format (role + content as string).
    return "openai"


# Valid framework identifiers for validation.
VALID_FRAMEWORKS = frozenset(
    {"openai", "anthropic", "agent_sdk", "adk", "langchain", "memosift", "vercel_ai"}
)
